de.dlr.sc.virsat.build/src/de/dlr/sc/virsat/build/validator/VirSatValidatorBuilder.java

```diff
@@ -152,41 +149,43 @@ protected void fullBuild(IProgressMonitor monitor, Set<IStructuralElementInstanc
 
 	}
 	
-	public static final String EXTENSION_POINT_ID_CONCEPT_VALIDATOR = "de.dlr.sc.virsat.build.DvlmValidator";
-	
 	/**
 	 * call this method to read the extension point for the concept images
 	 * It registers the found images in the bundles ImgageRegistry
 	 */
 	private void readConceptValidatorExtensionPoints() {
-		// Before reading the validators we remove all of them.
-		// By this the validators will not stack up anymore
 		seiValidators.clear();
 		repoValidators.clear();
 		
+		Set<String> activeConceptIds = getResourceSet().getRepository().getActiveConcepts().stream()
+				.map(c -> c.getFullQualifiedName()).collect(Collectors.toSet());
+		
 		// Now go to the registry and read them again. This is just a bug
 		// fix and may lead to performance issues.It should be considered
 		// to initialize them in a singleton manner later on
 		IExtensionRegistry registry = Platform.getExtensionRegistry();
 		
-		// Read all concept images through the extension point definitions in the various concept plugins
 		IConfigurationElement[] configElements = registry.getConfigurationElementsFor(EXTENSION_POINT_ID_CONCEPT_VALIDATOR);
 		
 		for (IConfigurationElement configElement : configElements) {
-			// String key = configElement.getAttribute("id");
-			try {
-				IConfigurationElement[] children = configElement.getChildren();
-				IConfigurationElement childConfigElement = children[0];
-				Object validator = childConfigElement.createExecutableExtension("class");
-				if (validator instanceof IStructuralElementInstanceValidator) {
-					IStructuralElementInstanceValidator seiValidator = (IStructuralElementInstanceValidator) validator;
-					seiValidators.add(seiValidator);
-				} else if (validator instanceof IRepositoryValidator) {
-					IRepositoryValidator repoValidator = (IRepositoryValidator) validator;
-					repoValidators.add(repoValidator);
+			for (IConfigurationElement validatorConfigElement : configElement.getChildren()) {
+				String contributor = validatorConfigElement.getContributor().getName();
+				boolean conceptSpecific = contributor.startsWith(CONCEPT_BUNDLE_PREFIX);
+				if (!conceptSpecific || activeConceptIds.contains(contributor)) {

```

Comment:
@andriikovalov-dlr Hey Andrii, I think we are introducing a lot of levels of nesting. Do you think we can reduce them somehow, or is it betetr to keep it like this?

Reply:
Yes, one level which we can remove in the future comes from the fact that our extension provides a collection of validators and not just one validator (hence nested for loop). I created a separate ticket for that #315.
On the cosmetic side, we can move validator creation (the `try` block) in a separate method.

Reply:
Extracted validator creation